crates/biome_js_analyze/src/semantic_analyzers/nursery/no_then_property.rs

```diff
@@ -0,0 +1,430 @@
+};
+use biome_js_syntax::{
+    AnyJsClassMember, AnyJsObjectMember, JsAssignmentExpression, JsCallExpression,
+    JsComputedMemberName, JsExport,
+};
+use biome_rowan::{declare_node_union, AstNode, AstSeparatedList, TextRange};
+
+declare_rule! {
+    /// Disallow `then` property.
+    ///
+    /// When combining objects with a `then` method (thenable objects) with await expressions or dynamic imports, caution is necessary.
+    /// These syntaxes interpret the object's then method as intended for the resolution or rejection of a promise, which can lead to unexpected behavior or errors.
+    ///
+    /// Source: https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-thenable.md
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// export {then};
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = {
+    ///     then() {}
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = {
+    ///     get then() {}
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const foo = {
+    ///    get then() {}
+    /// };
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.then = function () {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// class Foo {
+    ///     then() {}
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// class Foo {
+    ///     static then() {}
+    /// }
+    /// ```
+    ///
+    /// ## Valid
+    ///
+    /// ```js
+    /// export {then as success};
+    /// ```
+    ///
+    /// ```js
+    /// const foo = {
+    ///     success() {}
+    /// };
+    /// ```
+    ///
+    /// ```js
+    /// class Foo {
+    ///     success() {}
+    /// }
+    /// ```
+    ///
+    /// ```js
+    /// const foo = bar.then;
+    /// ```
+    ///
+    pub(crate) NoThenProperty {
+        version: "next",
+        name: "noThenProperty",
+        recommended: true,
+    }
+}
+
+declare_node_union! {
+    pub(crate) NoThenPropertyQuery =
+        AnyJsObjectMember |
+        JsComputedMemberName |
+        AnyJsClassMember |
+        JsAssignmentExpression |
+        JsCallExpression  |
+        JsExport
+}
+
+pub enum NoThenPropertyMessage {
+    Object,
+    Export,
+    Class,
+}
+
+pub struct RuleState {
+    range: TextRange,
+    message: NoThenPropertyMessage,
+}
+
+impl RuleState {
+    fn diagnostic_message(&self) -> MarkupBuf {
+        match self.message {
+            NoThenPropertyMessage::Object => {
+                markup! { "Do not add "<Emphasis>"`then`"</Emphasis>" to an object." }.to_owned()
+            }
+            NoThenPropertyMessage::Export => {
+                markup! { "Do not export "<Emphasis>"`then`"</Emphasis>"."}.to_owned()
+            }
+            NoThenPropertyMessage::Class => {
+                markup! {"Do not add "<Emphasis>"`then`"</Emphasis>" to a class." }.to_owned()
+            }
```

Comment:
It's enough to use Emphasis tag simply without back quote. This allows us to highlight the word.
```suggestion
            NoThenPropertyMessage::Object => {
                markup! { "Do not add "<Emphasis>"then"</Emphasis>" to an object." }.to_owned()
            }
            NoThenPropertyMessage::Export => {
                markup! { "Do not export "<Emphasis>"then"</Emphasis>"."}.to_owned()
            }
            NoThenPropertyMessage::Class => {
                markup! {"Do not add "<Emphasis>"then"</Emphasis>" to a class." }.to_owned()
            }
```